checks/raw/pinned_dependencies.go

```diff
@@ -61,9 +61,113 @@ func PinningDependencies(c *checker.CheckRequest) (checker.PinningDependenciesDa
 		return checker.PinningDependenciesData{}, err
 	}
 
+	if isStagedNugetDepsUnpinned(&results) {
+		if err := collectInsecureNugetCsproj(c, &results); err != nil {
+			return checker.PinningDependenciesData{}, err
+		}
+	} else {
+		promoteStagedNugetDependencies(&results, false)
+	}
```

Comment:
Do we need to have separate slices for staged vs. promoted deps? Can we just in-place mark nuget deps as pinned?


The code would roughly be updated to: (dont have my IDE in front of me so might have some syntax errors in here).
```go
if hasUnpinnedNuget(&results) && lockedProj(/* args */) {
    markNugetPinned(&results)
}

func markNugetPinned(r *results) {
   for i := range r.Dependencies {
      if isNuget(r.Dependencies[i]) {
          r.Dependencies[i].Pinned = asBoolPointer(true)
          r.Dependencies[i].Remediation = nil
      }
    }
}
```

Reply:
We considered having these on the same slice and your code makes sense for updating found nuget dependencies.
However, when we would need to remove nuget dependencies from the Dependencies slice and replacing them with new ones that are found during the post processing stage, the implementation seemed less clean.
an example for that is when some of the csproj files in a repo enable lock mode and some do not, we will not use the dependencies found during shell_validate_download, and instead build new ones at the post processing phase and stage them onto the Dependencies slice.
what are your thoughts?

Reply:
Currently the Pinned-Dependencies check collects the following:
1. Usages of unpinned dependencies in a file (e.g. mutable references for GitHub Actions `uses:` and Dockerfiles `FROM`)
2. Insecure shell patterns, which includes insecure package manager downloads and the nuget code.

> However, when we would need to remove nuget dependencies from the Dependencies slice and replacing them with new ones that are found during the post processing stage

My understanding was that the various install commands for `nuget`/`msbuild` could be secured in multiple ways:
* CLI flags (which are already detected by scorecard)
* this `RestorePackagesWithLockFile` property, which affects invocations of `msbuild`

So when we find a csproj file with the property set, I wasn't thinking of it as a new dependency to add to the slice, but rather a modification of an existing package manager download. This toggling is why I think this can be done in place, because we're not adding or removing anything, just marking pinned in some scenarios.

Reply:
our original idea was to provide more context. we changed the implementation so that the remediation string contains the csproj-level context of why the check failed, reverting the StagedDependencies collection.
Hope this version works better.

Reply:
> our original idea was to provide more context

This seems like a good goal thanks.